fix: Redirect user to OAuth client after first wrong login attempt

hugotiburtino opened the pull request:
~It has to be used in conjunction with remember=false.~ Actually, it doesn't, I was misled by the fact described [here](https://github.com/serlo/frontend/pull/1952#issuecomment-1345386501)

It should also fix the infinite login in serlo login page after coming from rocket chat.

Pushing flow id to url is a way of user not lose data. We may want to test if it is what we actually want to happen. edit: tested and it is indeed better, see discussion below.

Thinking: may we want to do the same for other flows... 🤔 edit: answer is yes

hugotiburtino committed:
fix(oauth): fix redirect to oauth client after first wrong pwd attempt

Has to be used in conjunction with remember=false in hydra.

Solves #1951

elbotho commented on src/components/pages/auth/login.tsx:
```diff
@@ -147,9 +147,9 @@ export function Login({ oauth }: { oauth?: boolean }) {
-    })
+
+    router.query.flow = flow.id
+    await router.push(router)
```
@hugotiburtino not sure why pushing router to router works and what it should do. Could you explain or send  a docs link?

hugotiburtino replied:
I was also surprised that it works. Based on https://stackoverflow.com/questions/61466891/nextjs-appending-a-query-param-to-a-dynamic-route.

Since push expects a Url and router has the corresponding fields it ends up working.

elbotho replied:
But what is the difference to the current code (besides the routing not being shallow)?

hugotiburtino replied:
It preserves the other query params that the url may have. In the specific case of the OAuth path it is important to maintain the login_challenge query parameter in order to solve the bug of the linked issue.

hugotiburtino replied:
We could alternatively not change the url at all and if the user refreshes the page they lose their previous inputs (and previous error messages,  if any). At the end of the day it is maybe what they want

elbotho replied:
thanks for the explanation.

>We could alternately not change the url at all and if the user refreshes the page they lose their previous inputs (and previous error messages, if any). At the end of the day it is maybe what they want

yeah, I think for login we should try that for now.
Maybe it's different for e.g. registering.

hugotiburtino replied:
Sorry, I have to correct myself: the recovery of data based on flow id is not implemented, Indeed we get the data from backend, but we don't fill in the fields with them So, the way it is now the user gets a very bad experience after refresh, 

See here in registration: error message with empty field
![image](https://user-images.githubusercontent.com/45924645/206875093-c81845f1-0841-4f18-8842-7dae9971d6df.png)

I think it is safe to delete these lines in login and in registration (I'll test with other flows and if I see that there isn't problem also you will get my commits on that)

hugotiburtino replied:
> I think it is safe to delete these lines in login and in registration (I'll test with other flows and if I see that there isn't problem also you will get my commits on that)

Of course, unless you think we should make effort to implement it now...

Otherwise, such lines are not only useless but also reason for bad UX. And there is no downside: if user reloads the page, they get another flow id and can go on.

hugotiburtino committed:
refactor: clean logging

hugotiburtino committed:
refactor(login): clean error messages after refresh

hugotiburtino committed:
refactor(login): avoid race between usual and oauth's redirect

Even when I was in a OAuth flow, my browser sometimes redirected me to
landing page.

hugotiburtino commented:
Regarding [086c9e6](https://github.com/serlo/frontend/pull/1952/commits/086c9e6b5ef2d7759c824ff1534b15c5bba182bb), I had the following bug: after clicking the Serlo login in local rocket chat, my browser redirected me sometimes to landing page, sometimes back to rocket chat. 
I supose there was some kind of race between the `oauthHandler` and `router.push(flow.return_to ?? redirection)`

hugotiburtino committed:
refactor(registration): clean error messages after refresh

hugotiburtino committed:
refactor(auth): fix endless redirect between login and recovery,

concentrating recovery url into a const.

hugotiburtino committed:
refactor(auth): clean messages after refresh from other flows

hugotiburtino committed:
refactor: remove unused import, minor refactoring and format

elbotho reviewed (approved):
👍

hugotiburtino commented:
Thanks for approving it. I'm not allowed to merge it. Could you do it?
